File: solutions/go/04/two_pointers/main.go
package main

import "fmt"

func countSubarrayLeK(xs []int, k int) int {
	n := len(xs)
	ans, left, sum := 0, 0, 0
	for right := 0; right < n; right++ {
		sum += xs[right]
		for sum > k && left <= right {
			sum -= xs[left]
			left++
		}
		ans += right - left + 1
	}
	return ans
}

func main() {
	fmt.Println(countSubarrayLeK([]int{1, 2, 3, 4}, 5))
}

File: solutions/python/04/two_pointers/main.py
def count_subarray_le_k(xs, k):
    n = len(xs)
    ans = left = s = 0
    for right in range(n):
        s += xs[right]
        while s > k and left <= right:
            s -= xs[left]
            left += 1
        ans += right - left + 1
    return ans


def main() -> None:
    print(count_subarray_le_k([1, 2, 3, 4], 5))


if __name__ == "__main__":
    main()
